File: 2023/golang/day2/day2.go
package day2

import (
	"strconv"
	"strings"
)

const MaxRed = 12
const MaxGreen = 13
const MaxBlue = 14

type Set struct {
	red, green, blue int
}

type Game struct {
	id   int
	sets []Set
}

type Challenge struct {
	games []Game
}

func Solve(input string, partTwo bool) int {
	c := newChallenge(input)

	if partTwo {
		return c.PartTwo()
	} else {
		return c.PartOne()
	}
}

func newChallenge(content string) *Challenge {
	c := &Challenge{}
	lines := strings.Split(content, "\n")
	for i, line := range lines {
		game := strings.Split(line, ":")
		id, _ := strconv.Atoi(game[0][5:])
		c.games = append(c.games, Game{})
		c.games[i].id = id
		sets := strings.Split(game[1], ";")
		for j, set := range sets {
			cubes := strings.Split(set, ",")
			c.games[i].sets = append(c.games[i].sets, Set{})
			c.games[i].sets[j].red = findCubeNumber(cubes, "red")
			c.games[i].sets[j].green = findCubeNumber(cubes, "green")
			c.games[i].sets[j].blue = findCubeNumber(cubes, "blue")
		}
	}
	return c
}

func (c Challenge) PartOne() int {
	var validGames []Game
	sum := 0
	for _, game := range c.games {
		if isValidGame(game) {
			validGames = append(validGames, game)
		}
	}

	for _, game := range validGames {
		sum += game.id
	}
	return sum
}

func (c Challenge) PartTwo() int {
	minSets := findMinSets(c.games)
	sum := 0
	for _, set := range minSets {
		product := set.red * set.green * set.blue
		sum += product
	}
	return sum
}

func findCubeNumber(cubes []string, color string) int {
	count := 0
	for _, cube := range cubes {
		cubeInfo := strings.Split(strings.TrimSpace(cube), " ")
		if strings.HasPrefix(cubeInfo[1], color) {
			count, _ = strconv.Atoi(cubeInfo[0])
		}
	}
	return count
}

func isValidGame(game Game) bool {
	isValid := true
	for _, set := range game.sets {
		if set.red > MaxRed || set.green > MaxGreen || set.blue > MaxBlue {
			isValid = false
			break
		}
	}
	return isValid
}

func findMinSets(games []Game) []Set {
	var minSets []Set
	for _, game := range games {
		minSet := Set{0, 0, 0}
		for _, set := range game.sets {
			if set.red > minSet.red {
				minSet.red = set.red
			}
			if set.green > minSet.green {
				minSet.green = set.green
			}
			if set.blue > minSet.blue {
				minSet.blue = set.blue
			}
		}
		minSets = append(minSets, minSet)
	}
	return minSets
}

File: 2023/typescript/src/challenges/day2.ts
import { Challenge } from './challenge';

interface Game {
  id: number;
  sets: Array<{
    red: number;
    green: number;
    blue: number;
  }>;
}

export class Day2 extends Challenge {
  private partTwo: boolean;
  private games: Array<Game>;
  private readonly maxRed = 12;
  private readonly maxGreen = 13;
  private readonly maxBlue = 14;

  constructor(input: string, partTwo: boolean) {
    super();
    this.games = input.split('\n').map((line) => {
      const game = line.split(':');
      return {
        id: parseInt(game[0]?.split(' ')[1] ?? '0'),
        sets: game[1].split(';').map((set) => {
          const cubes = set.split(',');
          return {
            red: parseInt(cubes.find((entry) => entry.includes('red'))?.replace(/\D/g, '') ?? '0'),
            green: parseInt(cubes.find((entry) => entry.includes('green'))?.replace(/\D/g, '') ?? '0'),
            blue: parseInt(cubes.find((entry) => entry.includes('blue'))?.replace(/\D/g, '') ?? '0'),
          };
        }),
      };
    });
    this.partTwo = partTwo;
  }

  solve = (): number => {
    return this.partTwo ? this.solveTwo() : this.solveOne();
  };

  private solveOne = (): number => {
    return this.games
      .filter((game) =>
        game.sets.every((set) => set.red <= this.maxRed && set.green <= this.maxGreen && set.blue <= this.maxBlue),
      )
      .reduce((sum, game) => sum + game.id, 0);
  };
  private solveTwo = (): number => {
    return this.games
      .map((game) => {
        const minSet = { red: 0, green: 0, blue: 0 };
        game.sets.forEach((set) => {
          minSet.red = set.red > minSet.red ? set.red : minSet.red;
          minSet.green = set.green > minSet.green ? set.green : minSet.green;
          minSet.blue = set.blue > minSet.blue ? set.blue : minSet.blue;
        });
        return minSet;
      })
      .map((set) => set.red * set.green * set.blue)
      .reduce((sum, product) => sum + product, 0);
  };
}
